### DISCUSSION TITLE
Meeting Start Time and Duration

### DISCUSSION THREAD

**Discussion by birdbm** [Jericho/ZoomNet#164] [Q&A] [answered]
I didn't want to enter this as an issue because I might be missing something.  The endpoint /users/{userId}/meetings (corresponding method = Meetings.GetAllAsync) returns an array of meetings as the records.  I'm curious why the StartTime and Duration properties were omitted from the Meeting class.  I saw these in the PastMeeting class but there doesn't appear to be any methods that return that object unless I'm missing something which I think I would have expected in the PastMeetings.GetAsync method.  I can also see these properties within the DashboardMetrics for meetings, but unfortunately am not able to use that feature because I am only using a Pro account.  Any guidance here would be appreciated.

**Comment by Jericho** [comment, marked as answer]
The abstract `Meeting` class contains properties that apply to all meeting types; any property that applies to only a specific type of meeting will be available on the specific meeting type class (i.e.: `RecurringMeeting`, `InstantMeeting` or `ScheduledMeeting`).

For example, `Password` and `CreatedOn` are two properties that apply to all meeting types which explains why they are available on the `Meeting` class but `RecurrenceInfo` only applies to recurring meetings which explains why it's available on the `RecurringMeeting` class and not on the `Meeting` class. All this to say that the two properties you mentioned are not available on the abstract `Meeting` class because they do not apply to all types of meetings, they only apply to scheduled meetings.

What this all means is that you must cast meetings to their concrete class in order to get access to properties that are specific to this type of meeting. You must do this whenever you fetch a single meeting (such as when you invoke the PastMeetings.GetAsync method, for example) or when you retrieve a list of meetings (such as when you invoke Meetings.GetAllAsync, for example).

For a code sample showing you how to this when retrieving a single meeting, please refer to this [question](https://github.com/Jericho/ZoomNet/issues/45). As for casting a list of records to their respective concrete class, here's how I do it with the help of the very handy `OfType<T>` Linq method:

```csharp
// Retrieve the first 100 meetings
var allMeetings = await client.Meetings.GetAllAsync(userId, MeetingListType.Scheduled, 100, null, cancellationToken).ConfigureAwait(false);

// Cast the meetings to their concrete class
var scheduledMeetings = allMeetings.Records.OfType<ScheduledMeeting>();
var instantMeetings = allMeetings.Records.OfType<InstantMeeting>();
var recurringMeetings = allMeetings.Records.OfType<RecurringMeeting>();
```

Let me know if this answers you question.

**Reply by birdbm** [reply]
Thank you so much! I thought I might be missing something but I didn't quite know what.  The fact that Meetings was an abstract class flew right over my head.  I didn't even realize you could cast objects on retrieval like that and still access the properties that weren't in the base class.  I really appreciate the time and thoroughness of the explanation.

**Comment by Jericho** [comment]
IMPORTANT UPDATE

ZoomNet version 0.44.0 will be released shortly and will introduce a breaking change that will render obsolete the C# code snippet I previously offered. The type of the data returned by the `GetAllAsync` method will no longer be the `Meeting` abstract class which you can cast into `ScheduledMeeting`, `InstantMeeting` or `RecurringMeeting` as appropriate.

It was brought to my attention (see [here](https://github.com/Jericho/ZoomNet/issues/223)) that when retrieving multiple meetings, the Zoom API returns summary information about the meetings and it does NOT include all information and therefore it does not make sense for ZoomNet to return `ScheduledMeeting`, `InstantMeeting` or `RecurringMeeting`. A new model class called `MeetingSummary` will be introduced and `GetAllAsync` will return an array of meeting summaries.

Here is a code snippet demonstrating how to display the start time for all meetings:
```csharp
// Retrieve the first 100 meetings
var allMeetings = await client.Meetings.GetAllAsync(userId, MeetingListType.Scheduled, 100, null, cancellationToken).ConfigureAwait(false);

// Display the start time for each meeting
foreach (var meeting in paginatedScheduledMeetings.Records)
{
    Debug.WriteLine($"Start date/time for meeting {meeting.Id} is {meeting.StartTime}");
}
```

**Reply by whampt** [reply]
This will be helpful for me working with recurring meetings and webinars. I'll do some testing to determine the best way to find the most recent instances for record-keeping purposes without having to use a webhook.
